feat: add configurable throttling mechanism for importing msgs

osmaczko opened the pull request:
part of: status-im/status-desktop#10815

Introduce slow and fast mode of messages import.
- slow mode: up to 10 msgs/sec
- fast mode: up to 1000 msgs/sec

It is up to the client to switch between these two modes, the default is slow mode.

In fast mode, the import time of a 4.2GB archive on my machine significantly dropped from around 2 hours to approximately 20 minutes, maintaining a stable CPU usage at about 5-7% without any spikes.
```
$ du -sh Status/data/archivedata 
4.2G    Status/data/archivedata
```

osmaczko committed:
feat: add configurable throttling mechanism for importing msgs

part of: status-im/status-desktop#10815

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (4)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 39ab25da | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3568/1/) | 2023-06-02 11:45:03 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230602-114231-39ab25-pr3568.zip) |
| :heavy_multiplication_x: | 39ab25da | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3568/1/) | 2023-06-02 11:45:25 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3568/1//consoleText) |
| :heavy_check_mark: | 39ab25da | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3568/1/) | 2023-06-02 11:46:17 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230602-114231-39ab25-pr3568.zip) |
| :heavy_check_mark: | 39ab25da | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3568/1/) | 2023-06-02 11:47:12 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230602-114231-39ab25-pr3568.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | c6c025ad | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3568/2/) | 2023-06-02 11:50:59 | ~57 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3568/2//consoleText) |
| :heavy_check_mark: | c6c025ad | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3568/2/) | 2023-06-02 11:52:36 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230602-115001-c6c025-pr3568.zip) |
| :heavy_check_mark: | c6c025ad | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3568/2/) | 2023-06-02 11:53:46 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230602-114956-c6c025-pr3568.aar) |
| :heavy_check_mark: | c6c025ad | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3568/2/) | 2023-06-02 11:54:59 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230602-114956-c6c025-pr3568.zip) |
| | | | | | | |
| :heavy_check_mark: | c596ed93 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3568/3/) | 2023-06-02 15:19:56 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230602-151743-c596ed-pr3568.zip) |
| :heavy_check_mark: | c596ed93 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3568/3/) | 2023-06-02 15:20:11 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230602-151745-c596ed-pr3568.zip) |
| :heavy_check_mark: | c596ed93 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3568/3/) | 2023-06-02 15:22:02 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230602-151743-c596ed-pr3568.aar) |
| :heavy_check_mark: | c596ed93 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3568/6/) | 2023-06-02 20:27:57 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3568/6//consoleText) |

jrainville reviewed (approved):
nice

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -47,6 +49,19 @@ var updateActiveMembersInterval = 24 * time.Hour
+
+const (
+	maxChunkSizeMessages = 1000
+	maxChunkSizeBytes    = 1500000
```
This one doesn't seem to be used any longer..

0x-r4bbit replied:
Actually, both are no longer used here

osmaczko replied:
It is still used when importing the community in `RequestImportDiscordCommunity` which is a bit different than importing messages from archives.

0x-r4bbit replied:
Ah sorry, forgot

0x-r4bbit commented on protocol/messenger_communities.go:
```diff
@@ -3637,3 +3676,42 @@ func (m *Messenger) CheckPermissionsToJoinCommunity(request *requests.CheckPermi
+	return chunks
+}
+
+func chunkAttachmentsByByteSize(slice []*protobuf.DiscordMessageAttachment, maxFileSizeBytes uint64) [][]*protobuf.DiscordMessageAttachment {
```
This function doesn't seem to be used anymore.

Is this not an issue with discord file attachments that are big?
One of the reasons this chunking was introduced was because we would run into scenarios where we'd process many messages that have discord message attachments, which would take a long time to save.

osmaczko replied:
It is still used when importing the community in `RequestImportDiscordCommunity` which is a bit different than importing messages from archives.

0x-r4bbit reviewed (commented):
Changes LGTM just a question:

Why exactly did we move the chunking before the message handling, instead of doing it inside the message handling?

Was that a reason for slow down?

osmaczko commented:
> Was that a reason for slow down?

The throttling mechanism we had implemented previously came with several drawbacks. We were handling a large volume of messages at once, often ~1k, before resting for 2 seconds. Processing these messages all at once frequently resulted in the database locking up during processing. New approach uses a rate limiter, leading to a more linear processing timeline, as it also takes into account the time of messages handling itself. Moreover, chunks are much smaller now, allowing for breathing room for other database operations more frequently.

> Why exactly did we move the chunking before the message handling, instead of doing it inside the message handling?

I moved the logic to the earliest possible point where we could break down the archive messages into smaller chunks to avoid logic duplication in separate handlers.

0x-r4bbit commented:
Thank you for clarifying @osmaczko !
